### PR TITLE
[Discover] Create Discover Shared plugin and features registry

### PR DESCRIPTION
**Pull request by tonyghiani**
## 📓 Summary

Part of https://github.com/elastic/kibana/issues/180024

This implementation aims to have a stateful layer that allows the management of dependencies between Discover and other plugins, reducing the need for a direct dependency. 

Although the initial thought was to have a plugin to register features for Discover, there might be other cases in future where we need to prevent cyclic dependencies. 
With this in mind, I created the plugin as a more generic solution to hold stateful logic as a communication layer for Discover <-> Plugins.

## Discover Shared

Based on some recurring naming in the Kibana codebase, `discover_shared` felt like the right place for owning these dependencies and exposing Discover functionalities to the external world.
It is initially pretty simple and only exposes a registry for the Discover features, but might be a good place to implement other upcoming concepts related to Discover.

Also, this plugin should ideally never depend on other solution plugins and keep its dependencies to a bare minimum of packages and core/data services.

```mermaid
flowchart TD

A(Discover) -- Get --> E[DiscoverShared]
B(Logs Explorer) -- Set --> E[DiscoverShared]
C(Security) -- Set --> E[DiscoverShared]
D(Any app) -- Set --> E[DiscoverShared]
```

## DiscoverFeaturesService

This service initializes and exposes a strictly typed registry to allow consumer apps to register additional features and Discover and retrieve them. 

The **README** file explains a real use case of when we'd need to use it and how to do that step-by-step.

Although it introduces a more nested folder structure, I decided to implement the service as a client-service and expose it through the plugin lifecycle methods to provide the necessary flexibility we might need:
- We don't know yet if any of the features we register will be done on the setup/start steps, so having the registry available in both places opens the door to any case we face.
- The service is client-only on purpose. My opinion is that if we ever need to register features such as server services or anything else, it should be scoped to a similar service dedicated for the server lifecycle and its environment.
It should never be possible to register the ObsAIAssistant presentational component from the server, as it should not be permitted to register a server service in the client registry.
A server DiscoverFeaturesService is not required yet for any feature, so I left it out to avoid overcomplicating the implementation.

## FeaturesRegistry

To have a strictly typed utility that suggests the available features on a registry and adheres to a base contract, the registry exposed on the DiscoverFeaturesService is an instance of the `FeaturesRegistry` class, which implements the registration/retrieval logic such that:
- If a feature is already registered, is not possible to override it and an error is thrown to notify the user of the collision.
- In case we need to react to registry changes, is possible to subscribe to the registry or obtain it as an observable for more complex scenarios.

The FeaturesRegistry already takes care of the required logic for the registry, so that `DiscoverFeaturesService`is left with the responsibility of instantiating/exposing an instance and provide the set of allowed features.

### CODE REVIEW

File: packages/kbn-discover-utils/src/services/features_registry/features_registry.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    this.features.next(new Map([...featuresMap.entries(), [feature.id, feature]]));
+  }
+
+  /**
+   * Retrieves all the registered features.
+   * @returns A Map instance.
+   */
+  getFeaturesMap() {
```
Comment (weltenwort) [inline packages/kbn-discover-utils/src/services/features_registry/features_registry.ts]: I'm concerned that this (and the observable/subscription below) returns a mutable map that can influence other consumers that have received the same instance. Do we need to expose it or would `register()` and `get()` be enough?

File: packages/kbn-discover-utils/src/services/features_registry/features_registry.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    return this.features.getValue();
+  }
+
+  /**
+   * Retrieves a registered feature by its id.
+   * @param id The identifier of the feature to retrieve.
+   * @returns The feature if found, otherwise undefined.
+   */
+  getById(id: Feature['id']) {
```
Comment (weltenwort) [inline packages/kbn-discover-utils/src/services/features_registry/features_registry.ts]: Can we try to narrow the return type?

File: plugins/.empty
Comment (weltenwort) [inline plugins/.empty:1]: Is this removal intentional?

File: src/plugins/discover_shared/README.md
Comment (weltenwort) [inline src/plugins/discover_shared/README.md:1]: Nice docs :heart:

File: src/plugins/discover_shared/public/services/discover_features/discover_features_service.ts
Comment (weltenwort) [inline src/plugins/discover_shared/public/services/discover_features/discover_features_service.ts:1]: Thanks for taking the time to cleanly implement it as a service!

File: x-pack/plugins/observability_solution/logs_explorer/public/components/common/translations.tsx [hunk 3]
```diff
@@ -227,8 +87,18 @@ export const contentHeaderTooltipParagraph1 = (
     id="xpack.logsExplorer.dataTable.header.content.tooltip.paragraph1"
     defaultMessage="Displays the document's {logLevel} and {message} fields."
     values={{
-      logLevel: <strong>log.level</strong>,
-      message: <strong>message</strong>,
+      logLevel: (
+        <strong>
+          {i18n.translate('xpack.logsExplorer..strong.loglevelLabel', {
```
Comment (weltenwort) [inline x-pack/plugins/observability_solution/logs_explorer/public/components/common/translations.tsx]: Should these really be translated? If they should, there's a duplicate `.` in the id.

**Review by weltenwort — COMMENTED** [review]
Looks nice overall, I just left a few questions below.

I'd like to defer judgement on the `discover_shared` plugin name to the DataDiscovery team.

File: packages/kbn-discover-utils/src/services/features_registry/features_registry.ts (hunk 1 above)
Comment (tonyghiani) [reply packages/kbn-discover-utils/src/services/features_registry/features_registry.ts]: It shouldn't be necessary at first, I'll make it private for its internal use only.

File: plugins/.empty
Comment (tonyghiani) [reply plugins/.empty:1]: Nope, I removed it by mistake when moving the new plugin, I'll restore it.

File: x-pack/plugins/observability_solution/logs_explorer/public/components/common/translations.tsx (hunk 3 above)
Comment (tonyghiani) [reply x-pack/plugins/observability_solution/logs_explorer/public/components/common/translations.tsx]: They shouldn't, I didn't event noticed that because the linter auto fix it due to the recent rule `@kbn/i18n/strings_should_be_translated_with_i18n`. I'll change it back.

File: packages/kbn-discover-utils/src/services/features_registry/features_registry.ts (hunk 1 above)
Comment (weltenwort) [reply packages/kbn-discover-utils/src/services/features_registry/features_registry.ts]: Thanks, but it still leaks out through `subscribe()` and `observe()`. Two consumers could receive the same `Map` and mutate it, thereby influencing each other.

Comment (tonyghiani) [reply packages/kbn-discover-utils/src/services/features_registry/features_registry.ts]: You are right, I didn't think about it. Let's keep it simple as we don't have requirements for this to be observed yet, I'll change the implementation to only expose the register and getById methods.

File: packages/kbn-discover-utils/index.ts [hunk 4]
```diff
@@ -44,3 +44,4 @@ export {
 } from './src';
 
 export * from './src/types';
+export * from './src/services/features_registry';
```
Comment (jughosta) [inline packages/kbn-discover-utils/index.ts]: Hi @tonyghiani,

Thanks for introducing a feature registry for Discover! It's a great way to prevent circular deps.

I would suggest to create a new package for this new service and name it the same way as the new plugin. Like `discover-shared` or `discover-registry`.

File: packages/kbn-discover-utils/src/services/features_registry/features_registry.ts (hunk 1 above)
Comment (weltenwort) [reply packages/kbn-discover-utils/src/services/features_registry/features_registry.ts]: That would solve it, thank you!

File: packages/kbn-discover-utils/index.ts (hunk 4 above)
Comment (tonyghiani) [reply packages/kbn-discover-utils/index.ts]: Hi @jughosta!

I considered this option as well and I've seen this pattern applied for the unified doc viewer package & plugin.
I created this on this discover utils package to avoid too many dependencies and since it's tree-shakeable, do you see any additional benefit in moving this utility into a different package?

P.S. Regarding the unified doc viewer package, it makes sense to have it in its package as it also holds the UI elements to create a more complex element and is consumed by different plugins, while this registry is ideally only used by discoverShared.

Comment (jughosta) [reply packages/kbn-discover-utils/index.ts]: For the separation of concerns it looks cleaner to me to have a separate package.

Comment (tonyghiani) [reply packages/kbn-discover-utils/index.ts]: I've implemented a new package for the registry in [5543214](https://github.com/tonyghiani/kibana/pull/1/commits/5543214b18e367dedb5ebf06403fa98c87719845), let me know if looks better now :)

File: .github/CODEOWNERS [hunk 5]
```diff
@@ -362,6 +362,8 @@ examples/developer_examples @elastic/appex-sharedux
 examples/discover_customization_examples @elastic/kibana-data-discovery
 x-pack/plugins/discover_enhanced @elastic/kibana-data-discovery
 src/plugins/discover @elastic/kibana-data-discovery
+src/plugins/discover_shared @elastic/kibana-data-discovery @elastic/obs-ux-logs-team
```
Comment (davismcphee) [inline .github/CODEOWNERS]: ++ to shared ownership 👍

**Review by davismcphee — APPROVED** [review]
Great work @tonyghiani! This seems like it will work well for our IoC needs in One Discover, and it's great that you included documentation too. Also, I think using `discover_shared` as the plugin name seems reasonable and clear. And with that said, this looks good to go on my end 👍

**Comment by tonyghiani** [comment]
After further offline discussion with @jughosta , we agreed that a new package for the FeaturesRegistry class would be unnecessary as it exposes nothing for other plugins except the newly created `discover-shared`. I moved its implementation into the plugin ([f53b567](https://github.com/tonyghiani/kibana/pull/1/commits/f53b567e13b79b13fdd2aa219d7f800c0e5a10be)) and removed the package to keep the implementation more concise.

**Review by weltenwort — APPROVED** [review]
LGTM - I've ignored anything specific to the fly-out code since I'll review that in #180262

**Comment by tonyghiani** [comment]
Thank you all, I'll reopen this PR against main once the other one will be merged to keep this implementation tracked on its own commit with some PR report, I might ask you an approval to merge it.

**Comment by tonyghiani** [comment]
Closing in favour of https://github.com/elastic/kibana/pull/181952
